File: dotnet/DesignPatterns/BehavioralDesignPatterns/StateDesignPattern/StateDesignPattern.Example1/ProcessingState.cs
namespace StateDesignPattern.Example1;

/// <summary>
/// Concrete State - represents an order being processed.
/// </summary>
public class ProcessingState : IOrderState
{
    public void ProcessOrder(Order order)
    {
        Console.WriteLine("Order is already being processed.");
    }

    public void ShipOrder(Order order)
    {
        Console.WriteLine("Order is being shipped...");
        order.SetState(new ShippedState());
        Console.WriteLine($"Order state changed to: {order.GetCurrentStateName()}");
    }

    public void DeliverOrder(Order order)
    {
        Console.WriteLine("Cannot deliver. Order must be shipped first.");
    }

    public void CancelOrder(Order order)
    {
        Console.WriteLine("Cancelling order during processing...");
        order.SetState(new CancelledState());
        Console.WriteLine($"Order state changed to: {order.GetCurrentStateName()}");
    }

    public string GetStateName() => "Processing";
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/StateDesignPattern/StateDesignPattern.Example1/src/ProcessingState.ts
import { IOrderState } from './IOrderState';
import { Order } from './Order';
import { ShippedState } from './ShippedState';
import { CancelledState } from './CancelledState';

export class ProcessingState implements IOrderState {
  processOrder(order: Order): void {
    console.log('Order is already being processed.');
  }

  shipOrder(order: Order): void {
    console.log('Order is being shipped...');
    order.setState(new ShippedState());
    console.log(`Order state changed to: ${order.getCurrentStateName()}`);
  }

  deliverOrder(order: Order): void {
    console.log('Cannot deliver. Order must be shipped first.');
  }

  cancelOrder(order: Order): void {
    console.log('Cancelling order during processing...');
    order.setState(new CancelledState());
    console.log(`Order state changed to: ${order.getCurrentStateName()}`);
  }

  getStateName(): string {
    return 'Processing';
  }
}
